### PR TITLE
Ravager Ignore Pain Ability is Now Endure

### PR DESCRIPTION
**Pull request by Surrealaser**
## About The Pull Request

Ignore Pain renamed to Endure, Cooldown decreased from 40 seconds to 30 seconds, and plasma cost increased from 175 to 200. It now has an aura SFX and additional notifications.

Ignore Pain/Endure now grants level 3 explosive resistance, immunity to the staggerslow effect from staggering munitions for its duration, and immunity to slowdown stacks and the stagger status in general.

Instead of granting effective immortality, Endure lets the Ravager remain active while at crit health and 100 additional crit health. If the Ravager ends Endure while in critical health, it will be restored to just above Critical Health.

Statistics tracking added for Endure.

## Why It's Good For The Game

Fixes a long standing bug?? with Ignore Pain that granted de facto immortality, while retaining its value and utility.

## Changelog
:cl:
fix: Instead of granting effective immortality, Endure lets the Ravager remain active while at crit health and 100 additional crit health. If the Ravager ends Endure while in critical health, it will be restored to just above Critical Health.
balance: Ignore Pain renamed to Endure, Cooldown decreased from 40 seconds to 30 seconds, and plasma cost increased from 175 to 200. It now has an aura SFX and additional notifications.
balance: Ignore Pain/Endure now grants level 3 explosive resistance, immunity to the staggerslow effect from staggering munitions for its duration, and immunity to slowdown stacks and the stagger status in general.
code: Statistics tracking added for Endure.
/:cl:

### CODE REVIEW

File: code/__DEFINES/traits.dm [hunk 1]
```diff
@@ -104,6 +105,8 @@
 #define TRAIT_UNDEFIBBABLE 		"undefibbable"//human can't be revived
 #define TRAIT_HEALING_INFUSION	"healing_infusion"//greatly improves natural healing for xenos
 #define TRAIT_PSY_DRAINED		"psy_drained"//mob was drained of life force by a xenos
+#define TRAIT_STAGGERIMMUNE		"stagger_immunity" //Immunity to stagger
+#define TRAIT_SLOWDOWNIMMUNE	"slowdown_immunity" //Immunity to slowdown
```
Comment (BraveMole) [inline code/__DEFINES/traits.dm:109]: this trait is never added, it's for later?

Comment (Surrealaser) [reply code/__DEFINES/traits.dm:109]: yeah

File: code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm [hunk 2]
```diff
@@ -139,50 +139,92 @@
...
 	X.visible_message("<span class='danger'>\The skin on the [X] begins to glow!</span>", \
 	"<span class='xenowarning'>We feel the plasma flowing through our veins!</span>")
 
-	X.ignore_pain = addtimer(VARSET_CALLBACK(X, ignore_pain, FALSE), 10 SECONDS)
-	X.ignore_pain_state = 0
-	addtimer(CALLBACK(GLOBAL_PROC, .proc/to_chat, X, "<span class='xenodanger'>We feel the plasma draining from our veins.</span>"), 7 SECONDS)
-	addtimer(CALLBACK(X, /mob/living/.proc/do_jitter_animation, 1000), 7 SECONDS)
+	X.endure = TRUE
+
+	X.add_filter("ravager_endure_outline", 4, list("type" = "outline", "size" = 1, "color" = COLOR_PURPLE)) //Set our cool aura; also confirmation we have the buff
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm:167]: we have new filter helpers for this

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm [hunk 3]
```diff
@@ -163,29 +163,30 @@
...
 
 /mob/living/carbon/xenomorph/adjustFireLoss(amount, updating_health = FALSE)
+
 	var/list/amount_mod = list()
 	SEND_SIGNAL(src, COMSIG_XENOMORPH_BURN_DAMAGE, amount, amount_mod)
 	for(var/i in amount_mod)
 		amount -= i
 
-	fireloss = clamp(fireloss + amount, 0, maxHealth - xeno_caste.crit_health)
+	fireloss = max(fireloss + amount, 0)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/damage_procs.dm:185]: why change this clamp?

File: code/modules/mob/living/living_health_procs.dm [hunk 4]
```diff
@@ -5,6 +5,12 @@
 /mob/living/proc/getBruteLoss()
 	return bruteloss
 
+///We straight up set bruteloss/brute damage to a desired amount
+/mob/living/proc/setBruteLoss(amount)
+	if(status_flags & GODMODE)
+		return FALSE	//godmode
```
Comment (TiviPlus) [inline code/modules/mob/living/living_health_procs.dm:11]:
```suggestion
///We straight up set bruteloss/brute damage to a desired amount unless godmode is enabled
/mob/living/proc/setBruteLoss(amount)
	if(status_flags & GODMODE)
		return FALSE
```

File: code/modules/mob/living/living_health_procs.dm [hunk 5]
```diff
@@ -16,6 +22,12 @@
 /mob/living/proc/getFireLoss()
 	return fireloss
 
+///We straight up set fireloss/burn damage to a desired amount
+/mob/living/proc/setFireLoss(amount)
+	if(status_flags & GODMODE)
+		return FALSE	//godmode
```
Comment (TiviPlus) [inline code/modules/mob/living/living_health_procs.dm:28]:
```suggestion
///We straight up set fireloss/burn damage to a desired amount unless godmode is enabled
/mob/living/proc/setFireLoss(amount)
	if(status_flags & GODMODE)
		return FALSE
```

File: code/modules/mob/living/status_procs.dm [hunk 6]
```diff
@@ -651,16 +656,27 @@
 ///This is where we normalize the set_slowdown input to be at least 0
 /mob/living/proc/adjust_slowdown(amount)
 	if(amount > 0)
+		if(HAS_TRAIT(src, TRAIT_SLOWDOWNIMMUNE)) //We're immune to slowdown
+			return slowdown
```
Comment (TiviPlus) [inline code/modules/mob/living/status_procs.dm:660]:
```suggestion
		if(HAS_TRAIT(src, TRAIT_SLOWDOWNIMMUNE))
			return slowdown
```
You shouldnt have comments unless they actually explain the code clearer

File: code/modules/mob/living/status_procs.dm [hunk 7]
```diff
@@ -651,16 +656,27 @@
...
 		set_slowdown(max(slowdown, amount)) //Slowdown overlaps rather than stacking.
 	else
 		set_slowdown(max(slowdown + amount, 0))
 	return slowdown
 
 /mob/living/proc/add_slowdown(amount, capped = 0)
+	if(HAS_TRAIT(src, TRAIT_SLOWDOWNIMMUNE)) //We're immune to slowdown
+		return
```
Comment (TiviPlus) [inline code/modules/mob/living/status_procs.dm:668]:
```suggestion
	if(HAS_TRAIT(src, TRAIT_SLOWDOWNIMMUNE))
		return
```

File: code/modules/mob/living/status_procs.dm [hunk 8]
```diff
@@ -651,16 +656,27 @@
...
 	if(slowdown)
 		adjust_slowdown(-STANDARD_SLOWDOWN_REGEN)
 	return slowdown
+
+/mob/living/carbon/xenomorph/add_slowdown(amount)
+	if(is_charging >= CHARGE_ON) //If we're charging we're immune to slowdown.
+		return
+	if(HAS_TRAIT(src, TRAIT_SLOWDOWNIMMUNE)) //We're immune to slowdown
+		return
+	adjust_slowdown(amount * XENO_SLOWDOWN_REGEN)
```
Comment (TiviPlus) [inline code/modules/mob/living/status_procs.dm:682]:
```suggestion
	if(is_charging >= CHARGE_ON)
		return
	if(HAS_TRAIT(src, TRAIT_SLOWDOWNIMMUNE))
		return
	adjust_slowdown(amount * XENO_SLOWDOWN_REGEN)
```

**Review by TiviPlus — COMMENTED** [review]
surface level review

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm (hunk 3 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/damage_procs.dm:185]: so Ravager can go into negative health properly

Comment (SplinterGP) [reply code/modules/mob/living/carbon/xenomorph/damage_procs.dm:185]: could probably replace by a STAT == "dead" which is what this was probably supposed to be and return

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/carbon_status_procs.dm [hunk 9]
```diff
@@ -8,6 +8,8 @@
 	losebreath = max(amount, 0)
 
 /mob/living/carbon/proc/set_stagger(amount)
+	if(amount > 0 && HAS_TRAIT(src, TRAIT_STAGGERIMMUNE)) //We're immune to stagger
+		return
 	stagger = max(amount, 0)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/carbon_status_procs.dm:13]:
```suggestion
	if(amount > 0 && HAS_TRAIT(src, TRAIT_STAGGERIMMUNE))
		return
	stagger = max(amount, 0)
```

File: code/__DEFINES/traits.dm [hunk 10]
```diff
@@ -87,6 +87,7 @@
 #define THROW_TRAIT "throw"
 #define FORTIFY_TRAIT "fortify" //Defender fortify ability.
 #define TRAIT_STASIS	"stasis"//Subject to the stasis effect
+#define ENDURE_TRAIT "trait" //Ravager Endure ability.
```
Comment (TiviPlus) [inline code/__DEFINES/traits.dm:90]:
```suggestion
#define ENDURE_TRAIT "endure" //Ravager Endure ability.
```

File: code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm [hunk 11]
```diff
@@ -139,50 +139,92 @@
...
+
+///Warns the player when Endure is about to end
+/datum/action/xeno_action/endure/proc/endure_warning()
+
+	to_chat(owner,"<span class='highdanger'>We feel the plasma draining from our veins... [ability_name] will last for only [RAVAGER_ENDURE_DURATION * (1-RAVAGER_ENDURE_WARNING) * 0.1] more seconds!</span>")
+	owner.playsound_local(owner, 'sound/voice/hiss4.ogg', 50, 0, 1)
+
+///Turns off the Endure buff
+/datum/action/xeno_action/endure/proc/endure_deactivate()
+	var/mob/living/carbon/xenomorph/ravager/R = owner
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm:191]: Should be typecast as xeno

File: code/modules/mob/living/status_procs.dm [hunk 12]
```diff
@@ -651,16 +656,27 @@
...
 
 ///Standard slowdown regen called by life.dm
 /mob/living/proc/handle_slowdown()
 	if(slowdown)
 		adjust_slowdown(-STANDARD_SLOWDOWN_REGEN)
 	return slowdown
+
+/mob/living/carbon/xenomorph/add_slowdown(amount)
+	if(is_charging >= CHARGE_ON)
+		return
```
Comment (TiviPlus) [inline code/modules/mob/living/status_procs.dm:679]: is this existing behavior and if no add it to CL+pr body /atomize it

File: code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm [hunk 13]
```diff
@@ -139,50 +139,92 @@
...
 
 	succeed_activate()
 	add_cooldown()
 
+	GLOB.round_statistics.ravager_endures++ //Statistics
+	SSblackbox.record_feedback("tally", "round_statistics", 1, "ravager_endures")
+
+///Warns the player when Endure is about to end
+/datum/action/xeno_action/endure/proc/endure_warning()
+
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm:184]:
```suggestion
/datum/action/xeno_action/endure/proc/endure_warning()
```

File: code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm [hunk 14]
```diff
@@ -139,50 +139,92 @@
...
+	to_chat(owner,"<span class='highdanger'>We feel the plasma draining from our veins... [ability_name] will last for only [RAVAGER_ENDURE_DURATION * (1-RAVAGER_ENDURE_WARNING) * 0.1] more seconds!</span>")
+	owner.playsound_local(owner, 'sound/voice/hiss4.ogg', 50, 0, 1)
+
+///Turns off the Endure buff
+/datum/action/xeno_action/endure/proc/endure_deactivate()
+	var/mob/living/carbon/xenomorph/R = owner
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/ravager/abilities_ravager.dm:190]: xenomorph should be X if we go by convention

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm [hunk 15]
```diff
@@ -163,29 +163,34 @@
...
+	if(stat == DEAD)
+		return
+
 	var/list/amount_mod = list()
 	SEND_SIGNAL(src, COMSIG_XENOMORPH_BURN_DAMAGE, amount, amount_mod)
 	for(var/i in amount_mod)
 		amount -= i
 
-	fireloss = clamp(fireloss + amount, 0, maxHealth - xeno_caste.crit_health)
+	fireloss = max(fireloss + amount, 0)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/damage_procs.dm:189]: Since you added the dead check you can undo this?

**Comment by TiviPlus** [comment]
Amount of damage you can tank needs to be decreased

I dont know why absolutely noone commented on the pr I hope you all know that not saying anything means this will just get merged without changes

**Comment by Naaanii** [comment]
https://cdn.discordapp.com/attachments/177287699043057665/848014327893590026/2021-05-29_02-22-10_Trim.mp4

**Comment by Losenis** [comment]
350hp, for context, is what you get at Ancient. Popping the ability essentially grants enough to be back to full health in practice, either for running away or just committing to the fight you spent that health at to finish it.

**Comment by Surrealaser** [comment]
> Amount of damage you can tank needs to be decreased
> 
> I dont know why absolutely noone commented on the pr I hope you all know that not saying anything means this will just get merged without changes

It's called fire.

If a Ravager takes lethal while in Endure, this is almost guaranteed to kill them if they have stacks when they come out of Endure.

**Comment by TiviPlus** [comment]
As said in discord change it to something like 200~

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**